File: be/src/main/java/be/movie36/service/LanguageService.java
package be.movie36.service;

import be.movie36.dto.request.LanguageRequest;
import be.movie36.dto.response.LanguageResponse;
import be.movie36.entity.Language;
import be.movie36.enums.Status;
import be.movie36.exception.AppException;
import be.movie36.exception.ErrorCode;
import be.movie36.repository.LanguageRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class LanguageService {
    private final LanguageRepository languageRepository;

    // tao moi
    public LanguageResponse create(LanguageRequest request) {
        if (languageRepository.findByName(request.getName()).isPresent()) {
            throw new AppException(ErrorCode.LANGUAGE_EXISTED);
        }
        Language language = Language.builder()
                .name(request.getName())
                .status(parseStatus(request.getStatus()))
                .build();
        return toResponse(languageRepository.save(language));
    }

    // lay ra tat ca
    public List<LanguageResponse> getAll() {
        return languageRepository.findAll().stream().map(this::toResponse).toList();
    }

    // lay tat ca dang o trang thai active
    public List<LanguageResponse> getAllActive() {
        return languageRepository.findByStatus(Status.ACTIVE).stream().map(this::toResponse).toList();
    }

    // lay theo id
    public LanguageResponse getById(Long id) {
        return toResponse(findById(id));
    }

    // cap nhat

    public LanguageResponse update(Long id, LanguageRequest request) {
        Language language = findById(id);

        // Kiểm tra tên mới có bị trùng với genre khác không
        languageRepository.findByName(request.getName())
                .ifPresent(existing -> {
                    if (!existing.getId().equals(id)) {
                        throw new AppException(ErrorCode.GENRE_EXISTED);
                    }
                });

        language.setName(request.getName());
        language.setStatus(parseStatus(request.getStatus()));

        return toResponse(languageRepository.save(language));
    }

    // xoa
    public void delete(Long id) {
        Language language = findById(id);
        language.setStatus(Status.INACTIVE);
        languageRepository.save(language);
    }

    // chuyen inactive sang active
    public  void restore(Long id){
        Language language = findById(id);
        language.setStatus(Status.ACTIVE);
        languageRepository.save(language);
    }

    // helper

    private Language findById(Long id) {
        return languageRepository.findById(id)
                .orElseThrow(() -> new AppException(ErrorCode.LANGUAGE_NOT_FOUND));
    }

    private Status parseStatus(String status) {
        if (status == null || status.isBlank()) {
            return Status.ACTIVE;
        }
        try {
            return Status.valueOf(status.toUpperCase());
        } catch (IllegalArgumentException e) {
            throw new AppException(ErrorCode.INVALID_STATUS);
        }
    }

    private LanguageResponse toResponse(Language language) {
        return LanguageResponse.builder()
                .id(language.getId())
                .name(language.getName())
                .status(language.getStatus().name())
                .build();
    }
}

File: fe/src/services/languageService.js
import axiosClient from './axiosClient';

const languageService = {
  // GET /api/language
  getAll: async (page = 0, size = 5, name = "") => {
    let url = `/language?page=${page}&size=${size}`;
    if (name) url += `&name=${encodeURIComponent(name)}`;
    const response = await axiosClient.get(url);
    return response.data;
  },

  // GET /api/language/active
  getAllActive: async () => {
    const response = await axiosClient.get('/language/active');
    return response.data;
  },

  // GET /api/language/:id
  getById: async (id) => {
    const response = await axiosClient.get(`/language/${id}`);
    return response.data;
  },

  // POST /api/language
  create: async (data) => {
    const response = await axiosClient.post(`/language`, data);
    return response.data;
  },

  // PUT /api/language/:id
  update: async (id, data) => {
    const response = await axiosClient.put(`/language/${id}`, data);
    return response.data;
  },

  // DELETE /api/language/:id
  delete: async (id) => {
    const response = await axiosClient.delete(`/language/${id}`);
    return response.data;
  },

  // PATCH /api/language/:id/restore
  restore: async (id) => {
    const response = await axiosClient.patch(`/language/${id}/restore`);
    return response.data;
  }
};

export default languageService;
